src/inspect/types.rs

```diff
@@ -0,0 +1,477 @@
+use std::fmt::{Display, Formatter};
+
+/// Designation of a Python type.
+///
+/// This enum is used to handle advanced types, such as types with generics.
+/// Its [`Display`] implementation can be used to convert to the type hint notation (e.g. `List[int]`).
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum TypeInfo {
+    /// The type `typing.Any`, which represents any possible value (unknown type).
+    Any,
+    /// The type `typing.None`.
+    None,
+    /// The type `typing.NoReturn`, which represents functions that never end (similar to `never` in Rust).
+    NoReturn,
+    /// The type `typing.Callable`.
+    ///
+    /// The first argument represents the parameters of the callable:
+    /// - `Some` of a vector of types to represent the signature,
+    /// - `None` if the signature is unknown (allows any number of arguments with type `Any`).
+    ///
+    /// The second argument represents the return type.
+    Callable(Option<Vec<TypeInfo>>, Box<TypeInfo>),
+    /// The type `typing.tuple`.
+    ///
+    /// The argument represents the contents of the tuple:
+    /// - `Some` of a vector of types to represent the accepted types,
+    /// - `Some` of an empty vector for the empty tuple,
+    /// - `None` if the number and type of accepted values is unknown.
+    ///
+    /// If the number of accepted values is unknown, but their type is, use [`Self::UnsizedTypedTuple`].
+    Tuple(Option<Vec<TypeInfo>>),
+    /// The type `typing.Tuple`.
+    ///
+    /// Use this variant to represent a tuple of unknown size but of known types.
+    ///
+    /// If the type is unknown, or if the number of elements is known, use [`Self::Tuple`].
+    UnsizedTypedTuple(Box<TypeInfo>),
+    /// A Python class.
+    Class {
+        /// The module this class comes from.
+        module: ModuleName,
+        /// The name of this class, as it appears in a type hint.
+        name: &'static str,
```

Comment:
Could someone want to create names at runtime? In that case `Cow<'static, str>` is more flexible 🤔

Reply:
Oh, I didn't know it was possible to create a `Cow` in a `const` context. That does sound more flexible.

Reply:
I don't really see a use case for creating names at runtime, but I don't really see a reason to forbid it either.